deployment/ccip/changeset/v1_6_2/cs_deploy_usdc_token_pools.go

```diff
@@ -0,0 +1,201 @@
+	}
+
+	// Check if a USDC token pool with the given version already exists
+	if _, ok := state.USDCTokenPools[deployment.Version1_6_2]; ok {
+		return fmt.Errorf("USDC token pool with version %s already exists on %s", deployment.Version1_6_2, chain)
+	}
+
+	// Perform USDC checks (i.e. make sure we can call the required functions)
+	// LocalMessageTransmitter and MessageBodyVersion are called in the contract constructor:
+	// https://github.com/smartcontractkit/chainlink/blob/f52a57762643b9cdc8e9241737e13501a4278716/contracts/src/v0.8/ccip/pools/USDC/USDCTokenPool.sol#L83
+	messenger, err := mock_usdc_token_messenger.NewMockE2EUSDCTokenMessenger(i.TokenMessenger, chain.Client)
+	if err != nil {
+		return fmt.Errorf("failed to connect address %s on %s with token messenger bindings: %w", i.TokenMessenger, chain, err)
+	}
+	_, err = messenger.LocalMessageTransmitter(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return fmt.Errorf("failed to fetch local message transmitter from address %s on %s: %w", i.TokenMessenger, chain, err)
+	}
+	_, err = messenger.MessageBodyVersion(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return fmt.Errorf("failed to fetch message body version from address %s on %s: %w", i.TokenMessenger, chain, err)
+	}
+
+	return nil
+}
+
+// DeployUSDCTokenPoolContractsConfig defines the USDC token pool contracts that need to be deployed on each chain.
+type DeployUSDCTokenPoolContractsConfig struct {
+	// USDCPools defines the per-chain configuration of each new USDC pool.
+	USDCPools    map[uint64]DeployUSDCTokenPoolInput
+	IsTestRouter bool
+}
+
+func deployUSDCTokenPoolContractsPrecondition(env cldf.Environment, c DeployUSDCTokenPoolContractsConfig) error {
+	state, err := stateview.LoadOnchainState(env)
+	if err != nil {
+		return fmt.Errorf("failed to load onchain state: %w", err)
+	}
+	for chainSelector, poolConfig := range c.USDCPools {
+		chain, chainState, err := state.GetEVMChainState(env, chainSelector)
+		if err != nil {
+			return fmt.Errorf("failed to get EVM chain state for chain selector %d: %w", chainSelector, err)
+		}
+		if !c.IsTestRouter && chainState.Router == nil {
+			return fmt.Errorf("missing router on %s", chain)
+		}
+		if c.IsTestRouter && chainState.TestRouter == nil {
+			return fmt.Errorf("missing test router on %s", chain)
+		}
+		err = poolConfig.Validate(env.GetContext(), chain, chainState)
+		if err != nil {
+			return fmt.Errorf("failed to validate USDC token pool config for chain selector %d: %w", chainSelector, err)
+		}
+	}
+	return nil
+}
+
+// DeployUSDCTokenPoolContractsChangeset deploys new USDC pools across multiple chains.
+func deployUSDCTokenPoolContractsLogic(env cldf.Environment, c DeployUSDCTokenPoolContractsConfig) (cldf.ChangesetOutput, error) {
+	if err := deployUSDCTokenPoolContractsPrecondition(env, c); err != nil {
+		return cldf.ChangesetOutput{}, fmt.Errorf("invalid DeployUSDCTokenPoolContractsConfig: %w", err)
+	}
+	newAddresses := cldf.NewMemoryAddressBook()
```

Comment:
```suggestion
	dataStore := NewMemoryDataStore()
```

Reply:
& then to write addresses:

```
  // Add an address reference to the datastore.
  err = dataStore.Addresses().Add(AddressRef{
    Address:         "0x123",
    ChainSelector:   1,
    Type:            "type1",
    Version:         semver.MustParse("1.0.0"),
    Qualifier:       "qualifier1",
    Labels:          NewLabelSet("label1"),
  })
  if err != nil {
    return err
  }
```

ref: http://docs.cld.cldev.sh/guides/changesets/datastore